### PR TITLE
Mining Updates

### LINKED ISSUE #12: Miner service to facilitate mining operations
**Description:**

The middleware will facilitate all communication between the miners and reward the appropriate miner on a first-come basis. 

- Initiate a docker container for the service
- Implement the proper endpoints required for the project
- Write unit tests and integration tests for all functions and endpoints

**Acceptance Criteria**
- [x] Mining works correctly, users are rewarded when mining works, a block is added to the blockchain when proofOfWork is computed

### PR DESCRIPTION
**Pull request by abhisachdev17**
## Summary of Changes
- Mining Pool Object which contains a pool of transactions
- One transaction is picked from the pool to be sent for Mining
- The transaction is only sent when the ready_to_mine condition variable with help of send is set.
- The ready_to_mine is set when the miners send back a valid proof of work.
TLDR; lots of OS stuff, let me know if you have any doubts on how things work.

Closes #12

### CODE REVIEW

File: src/backend/services/mining-api/src/MiningPool.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        self._mining_thread = threading.Thread(target=self.sendToMine, kwargs={'self': self, 'receiver': receiver})
+        self._mining_thread.daemon = True
+
+        self._send = threading.Condition()
+        self._ready_to_mine = ready_to_mine     
+
+    def addToPool(self, data):
+        self._pool.put(data)
+
+    def start_thread(self):
```
Comment (ZackHolmberg) [inline src/backend/services/mining-api/src/MiningPool.py:19]: I may be missing it but I don't see where this gets called. If that's the case, does it need to be called or can we remove it? Should be called in init I'd think

File: src/backend/services/transaction-api/src/routes.py [hunk 2]
```diff
@@ -13,6 +13,8 @@
 from shared import HttpCode, FailureReturnString
 
 
+ongoing_proof_for_id = None
```
Comment (ZackHolmberg) [inline src/backend/services/transaction-api/src/routes.py]: Is this supposed to be here?

File: src/backend/services/mining-api/src/routes.py [hunk 3]
```diff
@@ -1,13 +1,59 @@
...
+
+if os.environ.get('SERVICE_IN_DOCKER', False):
+    sys.path.append(os.path.abspath(os.path.join('..', '')))
+else:
+    sys.path.append(os.path.abspath(os.path.join('../..', '')))
+
+from shared import HttpCode, FailureReturnString
+
+
+ongoing_proof_for_id = None
```
Comment (ZackHolmberg) [inline src/backend/services/mining-api/src/routes.py]: I don't think this will be needed

File: src/backend/services/mining-api/src/routes.py [hunk 4]
```diff
@@ -1,13 +1,59 @@
...
+
 @socketio.on('echo')
 def handle(message):
-    socketio.emit("response", message, callback=msg)
+    socketio.emit("response", message)
+
+@socketio.on('proof')
+def handle_proofs(message):
+    transactions.ready_to_mine()
+    # TODO if proof for wrong id is received compare it with the current ongoing proof
```
Comment (ZackHolmberg) [inline src/backend/services/mining-api/src/routes.py:69]: Shouldn't be possible for a miner to send a proof for the wrong transaction as we will only have 1 mining session going at a time, and that mining will end server-side and client-side before another one begins

**Review by ZackHolmberg — CHANGES_REQUESTED** [review]

File: src/backend/.gitignore [hunk 5]
```diff
@@ -1,2 +1,3 @@
 **/env/
-**/__pycache__/
\ No newline at end of file
+**/__pycache__/
```
Comment (ZackHolmberg) [inline src/backend/.gitignore]:
```
**/.cache
**/**.pyc
tests/**.pyc
.pytest_cache
__init__.pyc
```

Add that to this file and then we don't have to have it in any of the service-level .gitignores

**Review by ZackHolmberg — APPROVED** [review]

**Review by akshaysharma21 — APPROVED** [review]
